### PR TITLE
u3d/*: failure reporter

### PR DESCRIPTION
**Pull request by niezbop**
### Pull Request Checklist

- [x] My pull request has been rebased on master
- [x] I ran `bundle exec rspec` to make sure that my PR didn't break any test
- [x] I ran `bundle exec rubocop` to make sure that my PR is inline with our code style
- [x] I have read the [code of conduct](https://github.com/DragonBox/u3d/blob/master/CODE_OF_CONDUCT.md)

### Pull Request Description

Addresses #146 
Requires #266 to be merged

This implements a simple failure report method, which stores failures so that they can be consulted later. The failure reporter is used in the log_analyzer.

### CODE REVIEW

File: lib/u3d/failure_reporter.rb [hunk 1]
```diff
@@ -0,0 +1,32 @@
+require 'date'
+require 'fileutils'
+require 'json'
+
+module U3d
+  class FailureReporter
+    class << self
+      def report(failure_type: "DEFAULT", failure_message: "", data: {})
+        return unless ENV['U3D_REPORT_FAILURES']
+        report = {
```
Comment (lacostej) [inline lib/u3d/failure_reporter.rb]: maybe wrap this into a rescue. We don't want this to fail the build in any way.

**Comment by lacostej** [comment]
Would logging the errors in the same place as the u3d.log file be a better idea? we wouldn't have to search for them. 

Have you looked at how to reuse the generated json files to improve the matchers?

**Comment by niezbop** [comment]
I'm not sure about `u3d.log`. Wouldn't this clutter the project? I believe that this can generate quite a lot of files so that's why I was keeping them in a separate place. Plus it would make cross-project comparison easier. Plus, they are mostly as hard to find as the default u3d.log I believe.

**Comment by lacostej** [comment]
My issue is that they would be on our build server, in a place where I can't really get a hold of them. Let's keep it like that for now. We might want to send them to a remote place later.

**Pull request merged**